### PR TITLE
Rewrite schema check

### PR DESCRIPTION
**Pull request by ds-lukaslangr**

### CODE REVIEW

File: src/datalakebundle/table/schema/MetadataChecker.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+from deepdiff import DeepDiff
+
+
+class MetadataChecker:
+    def __init__(self, logger: Logger, spark: SparkSession):
+        self.__logger = logger
+        self.__spark = spark
+
+    def check(self, table_definition: TableDefinition):
+        tables = self.__spark.sql(f"SHOW TABLES IN {table_definition.db_name}").toPandas()
```
Comment (kutny) [inline src/datalakebundle/table/schema/MetadataChecker.py]: Not sure but I expect that this would require pandas to be installed?

Comment (kutny) [reply src/datalakebundle/table/schema/MetadataChecker.py]: `self.__spark.sql(f"SHOW TABLES IN {table_definition.db_name}").collect()`

File: src/datalakebundle/table/schema/MetadataChecker.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+        if tables["tableName"].str.contains(table_definition.table_name).any():
+            self.__check_primary_key(table_definition)
+            self.__check_partition_by(table_definition)
+            self.__check_tbl_properties(table_definition)
+
+    def __check_primary_key(self, table_definition: TableDefinition):
+        pass
```
Comment (kutny) [inline src/datalakebundle/table/schema/MetadataChecker.py]: Why empty?

File: src/datalakebundle/table/schema/SchemaChecker.py [hunk 3]
```diff
@@ -1,51 +1,166 @@
...
+            if "iterable_item_removed" in ddiff:
+                result.extend(self.__get_iterable_item(expected_schema, ddiff["iterable_item_removed"], "missing field"))
+
+            if "dictionary_item_added" in ddiff:
+                result.extend(self.__get_dictionary_item(expected_schema, ddiff["dictionary_item_added"], "Unexpected field"))
+
+            if "dictionary_item_removed" in ddiff:
+                result.extend(self.__get_dictionary_item(expected_schema, ddiff["dictionary_item_removed"], "Missing field"))
+
+        print(result)
```
Comment (kutny) [inline src/datalakebundle/table/schema/SchemaChecker.py]: really print?

File: src/datalakebundle/table/schema/tests/SchemaChecker_test.py [hunk 4]
```diff
@@ -0,0 +1,378 @@
...
+
+class MyTestCase(unittest.TestCase):
+    def setUp(self) -> None:
+        self._container = bootstrapped_container.init("test")
+        logger = Logger("test")
+        mc: MetadataChecker = self._container.get(MetadataChecker)
+        self.schema_checker = SchemaChecker(logger, mc)
+
+    def test_no_diff(self):
+        assert self.schema_checker.generate_diff(df_schema, df_schema) == []
```
Comment (kutny) [inline src/datalakebundle/table/schema/tests/SchemaChecker_test.py]: self.assertEmptyArray() or any other self.assert* method should be used.

File: src/datalakebundle/table/schema/SchemaChecker.py [hunk 5]
```diff
@@ -30,7 +30,8 @@ def check(self, df: DataFrame, table_definition: TableDefinition):
             raise Exception(error_message)
 
     def check_metadata(self, table_definition: TableDefinition):
-        self.__metadata_checker.check(table_definition)
+        # self.__metadata_checker.check(table_definition)
+        pass
```
Comment (kutny) [inline src/datalakebundle/table/schema/SchemaChecker.py]: Why pass?

**Review by kutny — CHANGES_REQUESTED** [review]

File: src/datalakebundle/table/schema/MetadataChecker.py (hunk 2 above)
Comment (ds-lukaslangr) [reply src/datalakebundle/table/schema/MetadataChecker.py]: We haven't decided how we are going to check it. I just prepared the method for future. Should I delete it?

File: src/datalakebundle/table/schema/SchemaChecker.py (hunk 3 above)
Comment (ds-lukaslangr) [reply src/datalakebundle/table/schema/SchemaChecker.py]: already fixed

File: src/datalakebundle/table/schema/SchemaChecker.py (hunk 5 above)
Comment (ds-lukaslangr) [reply src/datalakebundle/table/schema/SchemaChecker.py]: We haven't decided if we wanted to turn it on. So it's disabled

File: src/datalakebundle/table/schema/tests/SchemaChecker_test.py (hunk 4 above)
Comment (ds-lukaslangr) [reply src/datalakebundle/table/schema/tests/SchemaChecker_test.py]: self.assertEmptyArray() doesn't exist so I used self.assertFalse() instead
Fixed the rest

File: src/datalakebundle/table/schema/MetadataChecker.py (hunk 1 above)
Comment (ds-lukaslangr) [reply src/datalakebundle/table/schema/MetadataChecker.py]: Fixed

File: src/datalakebundle/table/schema/MetadataChecker.py (hunk 2 above)
Comment (ds-lukaslangr) [reply src/datalakebundle/table/schema/MetadataChecker.py]: I have an idea how to check it. We could use the table_properties to store the primary_key

File: src/datalakebundle/table/schema/tests/SchemaChecker_test.py [hunk 6]
```diff
@@ -0,0 +1,393 @@
...
+                )
+            ),
+        ),
+        t.StructField("FIELD1", t.IntegerType()),
+        t.StructField("FIELD2", t.DoubleType()),
+    ],
+)
+
+
+class MyTestCase(unittest.TestCase):
```
Comment (kutny) [inline src/datalakebundle/table/schema/tests/SchemaChecker_test.py]: class MyTestCase, filename: SchemaChecker_test.py - why? Why not just SchemaCheckerTest(.py) ?

File: src/datalakebundle/table/schema/tests/SchemaChecker_test.py (hunk 4 above)
Comment (kutny) [reply src/datalakebundle/table/schema/tests/SchemaChecker_test.py]: sorry, there are no arrays in python, my PHP bad :-)

File: src/datalakebundle/table/schema/MetadataChecker.py (hunk 2 above)
Comment (kutny) [reply src/datalakebundle/table/schema/MetadataChecker.py]: If it is not implemented, let's remove the code. This solution does not make any sense.

File: src/datalakebundle/table/schema/MetadataChecker.py [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+
+class MetadataChecker:
+    def __init__(self, logger: Logger, spark: SparkSession):
+        self.__logger = logger
+        self.__spark = spark
+
+    def check(self, table_definition: TableDefinition):
+        tables = self.__spark.sql(f"SHOW TABLES IN {table_definition.db_name}")
+
+        if tables.filter(tables.tableName == table_definition.table_name).collect():
```
Comment (kutny) [inline src/datalakebundle/table/schema/MetadataChecker.py]: table_definition.table_name or table_definition.full_table_name? How to distinguish between equaly named tables stored in different databases?

File: src/datalakebundle/table/schema/MetadataChecker.py [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+            self.__check_tbl_properties(table_definition)
+
+    def __check_primary_key(self, table_definition: TableDefinition):
+        pass
+
+    def __check_partition_by(self, table_definition: TableDefinition):
+        try:
+            partitions_df = self.__spark.sql(f"SHOW PARTITIONS {table_definition.full_table_name}")
+        except AnalysisException:
+            self.__logger.info(f"Table {table_definition.full_table_name} is not partitioned.")
```
Comment (kutny) [inline src/datalakebundle/table/schema/MetadataChecker.py]: What is the point of this info message? Should we really send something to log if no partitioning exists? That is quite a normal situation...

File: src/datalakebundle/table/schema/MetadataChecker.py [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+                    extra["unexpected_keys"] = [v for v in unexpected_keys.values()]
+
+                if missing_keys:
+                    extra["missing_keys"] = [v for v in missing_keys.values()]
+
+                if values_changed:
+                    extra["values_changed"] = [f'{v["old_value"]} changed to {v["new_value"]}' for v in values_changed.values()]
+
+                self.__create_warning(
+                    "PARTITION_BY does NOT match schema",
```
Comment (kutny) [inline src/datalakebundle/table/schema/MetadataChecker.py]: TableSchema.partition_by does NOT match table schema?

Comment (kutny) [reply src/datalakebundle/table/schema/MetadataChecker.py]: What does this really mean? TablesSchema definies some partitions, that do not really exist in table? Or the other way arround? Or event something else?

File: src/datalakebundle/table/schema/MetadataChecker.py [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+                    extra,
+                    table_definition,
+                )
+
+    def __check_tbl_properties(self, table_definition: TableDefinition):
+        properties_df = self.__spark.sql(f"SHOW TBLPROPERTIES {table_definition.full_table_name}")
+        properties_pd = properties_df.toPandas().set_index("key")
```
Comment (kutny) [inline src/datalakebundle/table/schema/MetadataChecker.py]: Do we really need toPandas()?

File: src/datalakebundle/table/schema/MetadataChecker.py [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+        tbl_properties = properties_pd.to_dict()["value"]
+
+        tbl_properties.pop("Type", None)
+        tbl_properties.pop("delta.minReaderVersion", None)
+        tbl_properties.pop("delta.minWriterVersion", None)
```
Comment (kutny) [inline src/datalakebundle/table/schema/MetadataChecker.py]: Do we really need this hardcoded here?

File: src/datalakebundle/table/schema/tests/SchemaChecker_test.py [hunk 12]
```diff
@@ -0,0 +1,393 @@
...
+    def setUp(self) -> None:
+        self._container = bootstrapped_container.init("test")
+        logger = Logger("test")
+        mc: MetadataChecker = self._container.get(MetadataChecker)
+        self.schema_checker = SchemaChecker(logger, mc)
```
Comment (kutny) [inline src/datalakebundle/table/schema/tests/SchemaChecker_test.py]: Why not to load SchemaChecker from container?

File: src/datalakebundle/table/schema/tests/SchemaChecker_test.py [hunk 13]
```diff
@@ -0,0 +1,393 @@
...
+
+    def test_no_diff(self):
+        self.assertFalse(self.schema_checker.generate_diff(df_schema, df_schema))
```
Comment (kutny) [inline src/datalakebundle/table/schema/tests/SchemaChecker_test.py]: Why assertFalse() when self.schema_checker.generate_diff() returns list?

Comment (kutny) [reply src/datalakebundle/table/schema/tests/SchemaChecker_test.py]: What about self.assertListEqual([], ...)?

File: src/datalakebundle/table/schema/SchemaChecker.py [hunk 14]
```diff
@@ -1,51 +1,183 @@
...
-            extra = {
-                "table": full_table_name,
-            }
+            self.__logger.error(f"{error_message}\n{report}")
+            raise Exception(error_message)
 
-            if "set_item_removed" in ddiff:
-                extra["missing_fields"] = self.__pretty_ordered_set_to_output(ddiff["set_item_removed"])
+    def check_metadata(self, table_definition: TableDefinition):
+        # self.__metadata_checker.check(table_definition)
```
Comment (kutny) [inline src/datalakebundle/table/schema/SchemaChecker.py]: Why is this disabled?

File: src/datalakebundle/table/schema/SchemaChecker.py [hunk 15]
```diff
@@ -1,51 +1,183 @@
...
+            ch = chnks[0]
+            if len(ch) == 1 and len(chnks) == 1:
+                return
+            if ch[0] == "elementType":
+                yield "array"
+                yield from self.__rec_field_names(current["elementType"], chnks[1:])
+            else:
+                new = current[ch[0]][int(ch[1])]
+                yield new["name"]
+                yield from self.__rec_field_names(new["type"], chnks[1:])
```
Comment (kutny) [inline src/datalakebundle/table/schema/SchemaChecker.py]: What is "rec" in the method name?

Comment (kutny) [reply src/datalakebundle/table/schema/SchemaChecker.py]: The code in all the private methods bellow is very cryptic and does not have any tests.

**Review by kutny — CHANGES_REQUESTED** [review]

File: src/datalakebundle/table/schema/tests/SchemaChecker_test.py (hunk 6 above)
Comment (ds-lukaslangr) [reply src/datalakebundle/table/schema/tests/SchemaChecker_test.py]: fixed

File: src/datalakebundle/table/schema/MetadataChecker.py (hunk 7 above)
Comment (ds-lukaslangr) [reply src/datalakebundle/table/schema/MetadataChecker.py]: this line: tables = self.__spark.sql(f"SHOW TABLES IN {table_definition.db_name}") 
only gets tables from the correct database

File: src/datalakebundle/table/schema/MetadataChecker.py (hunk 9 above)
Comment (ds-lukaslangr) [reply src/datalakebundle/table/schema/MetadataChecker.py]: the `extra` dictionary explains where the exact difference is

File: src/datalakebundle/table/schema/MetadataChecker.py (hunk 10 above)
Comment (ds-lukaslangr) [reply src/datalakebundle/table/schema/MetadataChecker.py]: I think it's the simplest way to make a dict from a DataFrame
I can offer this instead `tbl_properties = {k:v for k,v in df.rdd.map(lambda x: (x.key, x.value)).collect()}`

File: src/datalakebundle/table/schema/tests/SchemaChecker_test.py (hunk 12 above)
Comment (ds-lukaslangr) [reply src/datalakebundle/table/schema/tests/SchemaChecker_test.py]: fixed

File: src/datalakebundle/table/schema/tests/SchemaChecker_test.py (hunk 13 above)
Comment (ds-lukaslangr) [reply src/datalakebundle/table/schema/tests/SchemaChecker_test.py]: fixed

File: src/datalakebundle/table/schema/MetadataChecker.py (hunk 8 above)
Comment (kutny) [reply src/datalakebundle/table/schema/MetadataChecker.py]: DESCRIBE TABLE --- and find partition info in the output?

File: src/datalakebundle/table/schema/MetadataChecker.py (hunk 11 above)
Comment (kutny) [reply src/datalakebundle/table/schema/MetadataChecker.py]: Please move this to bundle's YAML config

Comment (kutny) [reply src/datalakebundle/table/schema/MetadataChecker.py]: Dict of keys and values for tbl_properties

**Pull request merged**